### ISSUE TITLE
"InvalidCastException" on "add migration" after update to EF 3.0

### ISSUE THREAD

**Issue by ciacco85** [dotnet/efcore#18297]
After migration to all .net 3 stack, everything is working fine at runtime: startup, query ect
Adding a migration, i've this exception that I hadn't before with 2.2

System.InvalidCastException: Unable to cast object of type 'System.Boolean' to type 'System.Guid'.
   at lambda_method(Closure , ValueBuffer )
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalShadowEntityEntry..ctor(IStateManager stateManager, IEntityType entityType, ValueBuffer& valueBuffer)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntryFactory.NewInternalEntityEntry(IStateManager stateManager, IEntityType entityType, Object entity, ValueBuffer& valueBuffer)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntryFactory.Create(IStateManager stateManager, IEntityType entityType, Object entity, ValueBuffer& valueBuffer)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.CreateEntry(IDictionary`2 values, IEntityType entityType)
   at Microsoft.EntityFrameworkCore.Update.Internal.UpdateAdapter.CreateEntry(IDictionary`2 values, IEntityType entityType)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.TrackData(IModel source, IModel target)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.Diff(IModel source, IModel target, DiffContext diffContext)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationsModelDiffer.GetDifferences(IModel source, IModel target)
   at Microsoft.EntityFrameworkCore.Migrations.Design.MigrationsScaffolder.ScaffoldMigration(String migrationName, String rootNamespace, String subNamespace, String language)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.AddMigration(String name, String outputDir, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigrationImpl(String name, String outputDir, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigration.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.<>c__DisplayClass3_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
### Steps to reproduce
Since I don't know the entity involved and my DB has 60+ tables, I can't tell you reproduce step.
I've tried downloading the source code and reference it directly, but I couldn't manage to do it
What can I cant tell is that the last info message in --verbose mode is about seeding

'EFSDbContext' generated value '1' for the 'Discriminator' property of new 'Permission' entity.
Context 'EFSDbContext' started tracking 'Permission' entity with key '{Id: e86675e0-bd40-42e4-ac90-74d994d0d80b}'.
The 'Permission' entity with key '{Id: e86675e0-bd40-42e4-ac90-74d994d0d80b}' tracked by 'EFSDbContext' changed from 'Added' to 'Unchanged'.
'EFSDbContext' disposed.

I investigate in all seeding and I don't see any bool to guid mismatch (that i would have seen at compile time...)

<!--
What steps can we follow to reproduce the issue?

### Further technical details

EF Core version: 3.0.100
Database provider: (e.g. Microsoft.EntityFrameworkCore.SqlServer)
Target framework: (e.g. .NET Core 3.0)
Operating system:
IDE: (e.g. Visual Studio 2019 16.3 latest updated today

**Comment by ciacco85** [comment]
I'm using latest ef-core cli installed globally

**Comment by smitpatel** [comment]
@ciacco85 - It was reported earlier in #17851 but we are not able to get a repro code. It would really help us if you can provide a repro so that we can investigate. The error happens with seeding only. So perhaps entities using HasData are culprit. Also did you make any model changes between 2.2 -> 3.0 so that Add-Migration would add something?

**Comment by ciacco85** [comment]
No, I haven't made any changes: I need to Add an empty migration to add a store procedure.

Yes think so about seed, so I've removed the Seed part during endless attempts. I think the "Add migration" read all seed data and then, at the end, perform his comparison/tracking operation all together, so removing the seed didn't solve the problem.

I've attached a zip with the solution EFS.Database.sln, isolating the problem: I've removed unnecessary code, sensible data ect.
Repo step:
open solution, restore and build
open cdm/powershell inside EFS.Console folder
run "dotnet-ef database update"
run "dotnet-ef migrations add test" Invalid cast exception is thrown
run "dotnet-ef migrations add test --no-build --framework netcoreapp2.2" empty migration added without issue (removing the --no-build try to build the solution targeting 2.2 and so the build fails)
run " dotnet-ef migrations remove" error is thrown, but the migration is removed 
run " dotnet-ef migrations remove --no-build --framework netcoreapp2.2" migration removed without issue

database update after each migration are done without problem
[TestCode.zip](https://github.com/aspnet/EntityFrameworkCore/files/3711231/TestCode.zip)
